datashape/discovery.py

```
@@ -405,3 +405,29 @@ def descendents(d, x):
         children -= desc
         desc.update(children)
     return desc
+
+
+def _tp_eq(a, b):
+    return a == b or b is null
+
+
+@dispatch((range_, slice))
+def discover(r):
+    if isinstance(r, range_) and PY2:
+        raise NotImplementedError(
+            'xrange objects opaque and cannot be discovered'
+        )
+
+    types = map(discover, (r.start, r.stop, r.step))
+    for t in types:
+        if t is not null:
+            f = t
+            break
+    else:
+        # empty slice discovers to 'float64'
+        return Range(float64)
```

I think you should raise here instead of defaulting to `float64`